staging/src/k8s.io/client-go/dynamic/dynamicinformer/informer.go

```diff
@@ -135,7 +138,7 @@ func NewFilteredDynamicInformer(client dynamic.Interface, gvr schema.GroupVersio
 					return client.Resource(gvr).Namespace(namespace).Watch(context.TODO(), options)
 				},
 			},
-			&unstructured.Unstructured{},
+			&exampleObject,
```

Comment:
The documentation on shared index informer already states that exampleObject needs to have apiVersion and kind set for unstructured.Unstructured objects:

```
	// objectType is an example object of the type this informer is
	// expected to handle.  Only the type needs to be right, except
	// that when that is `unstructured.Unstructured` the object's
	// `"apiVersion"` and `"kind"` must also be right.
	objectType runtime.Object
```

If that was set, the expected apiVersion/kind would automatically get used for the description. Plumbing the group-version-resource in an annotation in the object is pretty weird...

Reply:
Yes, the docs do say that 😄  but it's impossible to do so from dynamic shared informers when the kind is not known and all we have instead is a GVR: https://github.com/kubernetes/kubernetes/blob/afebf498d7aa488097c8102adfcbc8abe7301ae4/staging/src/k8s.io/client-go/dynamic/dynamicinformer/informer.go#L120-L143

Reply:
hmm... rather than a magic annotation, what about a constructor that lets you set the description (and fixing the doc on the objectType fields that misleadingly says apiVersion/kind must be correct for unstructured objects)?

Reply:
Will look into doing that, thanks.

Reply:
@liggitt got some time to look into this. It will require updating `cache.NewSharedIndexInformer` or adding a new variant that takes functional options, as the description has to go from dynamic informer -> shared index informer -> cache.Config -> controller.Run -> NewReflector... is that ok?

Reply:
it's a decent chunk of plumbing, which is unfortunate, but I prefer that over magic annotations (especially since it would give an opportunity to correct the currently incorrect docs about apiVersion/kind being _required_ to be set in unstructured objects, and clarify how to set an alternate arbitrary description of the expected type)

Reply:
we're up to enough params and these are popular enough constructors that delegating existing constructors to a _WithOptions variant seems less disruptive and more usable